src/Shipping/ShippingZone.php

```diff
@@ -37,34 +48,223 @@ public function __construct( WC $wc ) {
-						break;
-					case 'state':
-						$country_code               = $this->get_country_of_state( $location->code );
-						$countries[ $country_code ] = $country_code;
-						break;
-					default:
-						break;
+			/**
+			 * @var WC_Shipping_Method[] $methods
+			 */
+			$methods = $zone->get_shipping_methods();
+
+			// Skip if no shipping methods.
+			if ( empty( $methods ) ) {
+				continue;
+			}
+
+			foreach ( $methods as $method ) {
+				// Check if the method is supported and return its properties.
+				$method = $this->parse_method( $method );
+
+				// Skip if method is not supported.
+				if ( null === $method ) {
+					continue;
+				}
+
+				// Add the method to the list of methods for each country in the zone.
+				foreach ( $this->get_shipping_countries_from_zone( $zone ) as $country ) {
+					// Initialize the shipping methods array if it doesn't exist.
+					$this->methods_countries[ $country ] = $this->methods_countries[ $country ] ?? [];
+
+					// Add the method to the array of shipping methods for the country if it doesn't exist.
+					if ( ! isset( $this->methods_countries[ $country ][ $method['id'] ] ) ||
+						 self::should_method_be_replaced( $method, $this->methods_countries[ $country ][ $method['id'] ] )
+					) {
+						$this->methods_countries[ $country ][ $method['id'] ] = $method;
+					}
 				}
 			}
 		}
+	}
 
-		// Match the list of shipping countries with the list of Merchant Center supported countries.
-		$countries = array_intersect( $countries, $this->get_mc_supported_countries() );
+	/**
+	 * Checks whether the shipping method should be replaced with a more suitable one.
+	 *
+	 * @param array $method
+	 * @param array $existing_method
+	 *
+	 * @return bool
+	 */
+	protected static function should_method_be_replaced( array $method, array $existing_method ): bool {
+		if ( $method['id'] !== $existing_method['id'] ) {
+			return false;
+		}
 
-		return array_values( $countries );
+		if (
+			// If a flat-rate/local-pickup method already exists, we replace it with the one with the higher cost.
+			self::does_method_have_higher_cost( $method, $existing_method ) ||
+			// If a free-shipping method already exists, we replace it with the one with the higher required minimum order amount.
+			self::does_method_have_higher_min_amount( $method, $existing_method )
+		) {
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks whether the given method has a higher cost than the existing method.
+	 *
+	 * @param array $method          A shipping method.
+	 * @param array $existing_method Another shipping method to compare with.
+	 *
+	 * @return bool
+	 */
+	protected static function does_method_have_higher_cost( array $method, array $existing_method ): bool {
+		if ( $method['id'] !== $existing_method['id'] ) {
+			return false;
+		}
+
+		return in_array( $method['id'], [ self::METHOD_FLAT_RATE, self::METHOD_PICKUP ], true ) && $method['options']['cost'] > $existing_method['options']['cost'];
```

Comment:
I thought we don't support pickup for now?

Reply:
Well, sort of... We still haven't decided whether to include them in this release or not. This is only in the service though and we won't return any suggested rates for the `pickup` method in our API. I just added it in there because it was very similar to the way the `flat-rate` method worked.